File: rust/src/equipment/item_creature.rs
use crate::core_mechanics::Attribute;
use crate::creatures::{Creature, CreatureCategory, HasModifiers, Modifier, calculate_minimum_level};

// Return a Creature that can be used to calculate item statistics for an item of the given rank.
// This standard creature has a power halfway between "minimum" and "maximum" reasonable power:
// R1: 2
// R2: 4
// R3: 6
// R4: 8
// R5: 10
// R6: 12
// R7: 14
pub fn item_creature(rank: i32) -> Creature {
    let level = calculate_minimum_level(rank);

    // Start with a 1 perception and increase it at each attribute scaling level.
    // This is somewhere between minimum and maximum reasonable accuracy.
    let perception = 1 + ((level + 3) / 6);

    let power_attribute = match rank {
        -1 => 0,
        0 => 0,
        1 => 2,  // target: 2
        2 => 2,  // target: 4
        3 => 3,  // target: 6
        4 => 3,  // target: 8
        5 => 4,  // target: 10
        6 => 4,  // target: 12
        7 => 5,  // target: 14
        8 => 6, // target: 16
        _ => panic!("Unsupported rank {}", rank),
    };

    let mut creature = Creature::new(level, CreatureCategory::Character);
    creature.add_modifier(
        Modifier::Attribute(Attribute::Perception, perception),
        None,
        None,
    );
    creature.add_modifier(
        Modifier::Attribute(Attribute::Strength, power_attribute),
        None,
        None,
    );
    creature.add_modifier(
        Modifier::Attribute(Attribute::Willpower, power_attribute),
        None,
        None,
    );

    return creature;
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::core_mechanics::attacks::HasAttacks;

    #[test]
    fn has_correct_power_at_rank() {
        for rank in 0..8 {
            assert_eq!(
                rank * 2,
                item_creature(rank).calc_power(false),
                "Should have correct power at rank {}; {}",
                rank,
                item_creature(rank).to_latex(),
            );
        }
    }
}

File: typescript/src/equipment/item_creature.ts
import { Creature } from '../character_sheet/creature';

export function calculateMinimumLevel(rank: number): number {
  switch (rank) {
    case -1:
      return 0;
    case 0:
      return 0;
    case 1:
      return 1;
    case 2:
      return 4;
    case 3:
      return 7;
    case 4:
      return 10;
    case 5:
      return 13;
    case 6:
      return 16;
    case 7:
      return 19;
    case 8:
      return 21;
    default:
      throw new Error(`Unsupported rank ${rank}`);
  }
}

export function getItemCreature(rank: number): Creature {
  const level = calculateMinimumLevel(rank);

  // Start with a 2 perception and increase it at each attribute scaling level.
  const perception = 2 + Math.floor((level + 3) / 6);

  let powerAttribute: number;
  switch (rank) {
    case -1:
    case 0:
      powerAttribute = 0;
      break;
    case 1:
    case 2:
      powerAttribute = 2;
      break;
    case 3:
    case 4:
      powerAttribute = 3;
      break;
    case 5:
    case 6:
      powerAttribute = 4;
      break;
    case 7:
      powerAttribute = 5;
      break;
    case 8:
      powerAttribute = 6;
      break;
    default:
      throw new Error(`Unsupported rank ${rank}`);
  }

  const creature = Creature.new();
  creature.setRequiredProperties({
    alignment: 'neutral',
    base_class: 'hybrid',
    elite: false,
    creature_origin: 'natural',
    creature_type: 'humanoid',
    size: 'medium',
    level: level,
  });
  creature.setProperties({ name: 'The Item' });

  creature.addSimpleModifier({
    name: 'Item Scaling Perception',
    statistic: 'perception',
    value: perception,
  });
  creature.addSimpleModifier({
    name: 'Item Scaling Strength',
    statistic: 'strength',
    value: powerAttribute,
  });
  creature.addSimpleModifier({
    name: 'Item Scaling Willpower',
    statistic: 'willpower',
    value: powerAttribute,
  });

  return creature;
}
